File: algorithms/460.LFU-Cache.py
#!/usr/bin/env python3
from collections import defaultdict, deque

# Average O(1) algorithm for put() & get()
# but for duplicate keys, it can be slow because of deque's remove - O(n)
# We can improve this using real linked-list instead of using deque, double-linked list of array.
class LFUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.key_to_val = {}
        self.key_to_freq = defaultdict(int)
        self.freq_to_keys = defaultdict(deque)
        self.min_freq = 0
        
    def get(self, key:int) -> int:
        if key not in self.key_to_val:
            return -1
            
        # Increase frequency
        old_freq = self.key_to_freq[key]
        self.key_to_freq[key] += 1
        new_freq = self.key_to_freq[key]
        
        # Update frequency map
        self.freq_to_keys[old_freq].remove(key)
        self.freq_to_keys[new_freq].append(key)
        
        # If the previous min_freq bucket is empty, update min_freq
        if not self.freq_to_keys[old_freq]:
            del self.freq_to_keys[old_freq]
            if self.min_freq == old_freq:
                self.min_freq += 1
            
        return self.key_to_val[key]
        
    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return
        
        if key in self.key_to_val:
            # Update existing key_to_val & increase the frequency
            self.key_to_val[key] = value    
            self.get(key) 
        else:
            # Insert new key
            if len(self.key_to_val) >= self.capacity:
                # Evict LFU: least frequently used item
                lfu_key = self.freq_to_keys[self.min_freq].popleft()
                del self.key_to_val[lfu_key]
                del self.key_to_freq[lfu_key]

            # Insert new key with initial frequency
            self.key_to_val[key] = value
            self.key_to_freq[key] = 1
            self.freq_to_keys[1].append(key)
            self.min_freq = 1

File: algorithms/460.LFU-Cache.java
import java.util.*;

public class LFUCache {
	private final int capacity;
	private int minFreq;
	private final Map<Integer, Integer> keyToVal;
	private final Map<Integer, Integer> keyToFreq;
	private final Map<Integer, LinkedHashSet<Integer>> freqToKeys;

	public LFUCache(int capacity) {
		this.capacity = capacity;
		this.minFreq = 0;
		this.keyToVal = new HashMap<>();
		this.keyToFreq = new HashMap<>();
		this.freqToKeys = new HashMap<>();
	}

	public int get(int key) {
		if (!keyToVal.containsKey(key)) {
			return -1;
		}

		// Increase frequency
		int oldFreq = keyToFreq.get(key);
		int newFreq = oldFreq + 1;
		keyToFreq.put(key, newFreq);

		// Update frequency map
		freqToKeys.get(oldFreq).remove(key);
		freqToKeys.computeIfAbsent(newFreq, k -> new LinkedHashSet<>()).add(key);

		// If the previous minFreq bucket is empty, update minFreq
		if (freqToKeys.get(oldFreq).isEmpty()) {
			freqToKeys.remove(oldFreq);
			if (minFreq == oldFreq) {
				minFreq++;
			}
		}

		return keyToVal.get(key);
	}

	public void put(int key, int value) {
		if (capacity == 0) {
			return;
		}

		if (keyToVal.containsKey(key)) {
			// Update existing key's value and increase the frequency
			keyToVal.put(key, value);
			get(key);
			return;
		}

		// Insert new key
		if (keyToVal.size() >= capacity) {
			// Evict LFU: least frequently used item
			int lfuKey = freqToKeys.get(minFreq).iterator().next();
			freqToKeys.get(minFreq).remove(lfuKey);
			if (freqToKeys.get(minFreq).isEmpty()) {
				freqToKeys.remove(minFreq);
			}
			keyToVal.remove(lfuKey);
			keyToFreq.remove(lfuKey);
		}

		// Insert new key with initial frequency
		keyToVal.put(key, value);
		keyToFreq.put(key, 1);
		freqToKeys.computeIfAbsent(1, k -> new LinkedHashSet<>()).add(key);
		minFreq = 1;
	}

	// public static void main(String[] args) {
	//	 LFUCache cache = new LFUCache(2);
	//	 cache.put(1, 1);
	//	 cache.put(2, 2);
	//	 System.out.println(cache.get(1)); // returns 1
	//	 cache.put(3, 3); // evicts key 2
	//	 System.out.println(cache.get(2)); // returns -1 (not found)
	//	 System.out.println(cache.get(3)); // returns 3
	//	 cache.put(4, 4); // evicts key 1
	//	 System.out.println(cache.get(1)); // returns -1 (not found)
	//	 System.out.println(cache.get(3)); // returns 3
	//	 System.out.println(cache.get(4)); // returns 4
	// }
}
